File: packages/sdk-ts/src/caps/memory.ts
/**
 * In-process caps backend. **For tests only.**
 *
 * Counters live in one process's memory: they are not shared between workers,
 * are lost on restart, and would let a multi-process deployment exceed every
 * cap. Use the Redis or Postgres backend in any deployment.
 */

import { CapExceededError, WINDOW_MS, tenantHash, type CapsBackend, type ResolvedCapLimit } from './meter.js';

interface Entry {
  at: number;
  reservationId: string;
  units: number;
}

export class InMemoryCapsBackend implements CapsBackend {
  readonly #counters = new Map<string, Entry[]>();

  async #entries(tenantId: string, limit: ResolvedCapLimit, now: number): Promise<Entry[]> {
    const id = `${await tenantHash(tenantId)}:${limit.key}`;
    let entries = this.#counters.get(id);
    if (entries === undefined) {
      entries = [];
      this.#counters.set(id, entries);
    }
    const window = WINDOW_MS[limit.window];
    if (window > 0) {
      const live = entries.filter((e) => e.at > now - window);
      entries.splice(0, entries.length, ...live);
    }
    return entries;
  }

  async reserve(tenantId: string, reservationId: string, limits: readonly ResolvedCapLimit[], nowMs: number | undefined): Promise<void> {
    const now = nowMs ?? Date.now();
    // Resolve every counter before checking, then check and record with no
    // await in between: JavaScript runs this block without interleaving.
    const all: Entry[][] = [];
    for (const limit of limits) all.push(await this.#entries(tenantId, limit, now));
    limits.forEach((limit, i) => {
      const used = (all[i] as Entry[]).reduce((sum, e) => sum + e.units, 0);
      if (used + limit.units > limit.limit) {
        throw new CapExceededError({
          limit: limit.limit, window: limit.window, used, requested: limit.units, scope: limit.scope, kind: limit.kind,
        });
      }
    });
    limits.forEach((limit, i) => {
      (all[i] as Entry[]).push({ at: now, reservationId, units: limit.units });
    });
  }

  async refund(tenantId: string, reservationId: string, limits: readonly ResolvedCapLimit[]): Promise<void> {
    for (const limit of limits) {
      const entries = this.#counters.get(`${await tenantHash(tenantId)}:${limit.key}`);
      if (entries === undefined) continue;
      const kept = entries.filter((e) => e.reservationId !== reservationId);
      entries.splice(0, entries.length, ...kept);
    }
  }

  async usage(tenantId: string, limit: ResolvedCapLimit, nowMs: number | undefined): Promise<number> {
    return (await this.#entries(tenantId, limit, nowMs ?? Date.now())).reduce((sum, e) => sum + e.units, 0);
  }
}

File: packages/sdk-py/src/grantex/caps/_memory.py
"""In-process caps backend. **For tests only.**

Counters live in one process's memory: they are not shared between workers,
are lost on restart, and would let a multi-process deployment exceed every
cap. Use the Redis or Postgres backend in any deployment.
"""

from __future__ import annotations

import threading
import time
from typing import Dict, List, Optional, Sequence, Tuple

from ._meter import WINDOW_MS, CapExceededError, CapLimit, tenant_hash

# (tenant hash, counter key) -> list of (reserved_at_ms, reservation id, units)
_Entries = List[Tuple[int, str, int]]


class InMemoryCapsBackend:
    """Thread-safe, single-process counters for unit tests."""

    def __init__(self) -> None:
        self._lock = threading.Lock()
        self._counters: Dict[Tuple[str, str], _Entries] = {}

    @staticmethod
    def _now(now_ms: Optional[int]) -> int:
        return now_ms if now_ms is not None else int(time.time() * 1000)

    def _live(self, tenant_id: str, limit: CapLimit, now: int) -> _Entries:
        entries = self._counters.setdefault((tenant_hash(tenant_id), limit.key()), [])
        window = WINDOW_MS[limit.window]
        if window > 0:
            entries[:] = [e for e in entries if e[0] > now - window]
        return entries

    def reserve(
        self,
        tenant_id: str,
        reservation_id: str,
        limits: Sequence[CapLimit],
        now_ms: Optional[int],
    ) -> None:
        now = self._now(now_ms)
        with self._lock:
            live = [self._live(tenant_id, limit, now) for limit in limits]
            for limit, entries in zip(limits, live):
                used = sum(units for _, _, units in entries)
                if used + limit.units > limit.limit:
                    raise CapExceededError(
                        limit=limit.limit, window=limit.window, used=used,
                        requested=limit.units, scope=limit.scope, kind=limit.kind,
                    )
            for limit, entries in zip(limits, live):
                entries.append((now, reservation_id, limit.units))

    def refund(
        self,
        tenant_id: str,
        reservation_id: str,
        limits: Sequence[CapLimit],
        now_ms: Optional[int],
    ) -> None:
        with self._lock:
            for limit in limits:
                entries = self._counters.get((tenant_hash(tenant_id), limit.key()), [])
                entries[:] = [e for e in entries if e[1] != reservation_id]

    def usage(self, tenant_id: str, limit: CapLimit, now_ms: Optional[int]) -> int:
        with self._lock:
            return sum(units for _, _, units in self._live(tenant_id, limit, self._now(now_ms)))
